informer/informer.go

```diff
@@ -0,0 +1,383 @@
+// package informer provides primitives to watch event objects from the
+// Kubernetes API in a deterministic way. The following conditions are
+// guaranteed by the watcher.
+//
+//     - The informer is able to watch all kinds of objects as soon as a proper
+//       watch endpoint and a factory implementing ZeroObjectFactory is given.
+//     - Events for objects that are created, deleted or updated are dispatched
+//       immeditiatelly.
+//     - Events for objects that are created or updated are dispatched via the
+//       same channel. The informer cannot distinguish between a created or
+//       updated event object.
+//     - Events for objects that are alredy cached are not dispatched during the
+//       configured resync period, but periodically after it.
+//     - Events for objects are never dispatched twice.
+//     - Events for objects can be dispatched in a rate limitted way, if
+//       configured accordingly.
+//
+package informer
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"github.com/cenk/backoff"
+	"github.com/giantswarm/microerror"
+	"k8s.io/apimachinery/pkg/watch"
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/tools/cache"
+)
+
+const (
+	// DefaultRateWait is the default value for the RateWait setting. See Config
+	// for more information.
+	DefaultRateWait = 0 * time.Second
+	// DefaultResyncPeriod is the default value for the ResyncPeriod setting. See
+	// Config for more information.
+	DefaultResyncPeriod = 1 * time.Minute
+)
+
+// Config represents the configuration used to create a new Informer.
+type Config struct {
+	// Dependencies.
+	BackOff    backoff.BackOff
+	RestClient rest.Interface
+
+	// Settings.
+
+	// RateWait provides configuration for some kind of rate limitting. The
+	// informer watch provides events via the update channel every ResyncPeriod.
+	// This triggers the release of update events. RateWait is the time to wait
+	// between released events.
+	RateWait time.Duration
+	// ResyncPeriod is the time to wait before releasing update events again.
+	ResyncPeriod time.Duration
+}
+
+// DefaultConfig provides a default configuration to create a new by best
+// effort.
+func DefaultConfig() Config {
+	return Config{
+		// Dependencies.
+		BackOff:    nil,
+		RestClient: nil,
+
+		// Settings.
+		RateWait:     DefaultRateWait,
+		ResyncPeriod: DefaultResyncPeriod,
+	}
+}
+
+// Informer implements primitives to watch event objects from the Kubernetes API
+// in a deterministic way.
+type Informer struct {
+	// Dependencies.
+	backOff    backoff.BackOff
+	restClient rest.Interface
+
+	// Internals.
+	cache       *sync.Map
+	initializer chan struct{}
+
+	// Settings.
+	rateWait     time.Duration
+	resyncPeriod time.Duration
+}
+
+// New creates a new Informer.
+func New(config Config) (*Informer, error) {
```

Comment:
Here is the new informer. I did intensive manual testing. I still have to write unit tests but I would be fine with doing this in the next PRs. On demand I can provide demos and scripts for manual testing.